Add the ability to push history to a remote Project Haystack server

sjlongland opened the pull request:
The following branch adds support for pushing history to a remote Project Haystack server (e.g. another nHaystack system, SkySpark or WideSky).  It also supports creation of `point` entities on the remote server in situations where the server implements a compatible `createRec` operation.

It has been successfully tested with both nHaystack and WideSky.

The branch is based on a merge with the `n4` branch; so includes all that branch's changes -- I can cherry-pick this directly onto `n4-development` if required.

sjlongland committed:
[rt] Add test stations for history upload.

`nhaystack_sender` is a JACE that pushes its history to another
nHaystack instance running on another JACE via `hisWrite` calls.

`nhaystack_receiver` is a JACE that receives the history from
`nhaystack_sender`.

There's no corresponding test in the `nhaystack` tests, rather the test
case is you spin both up, go away for a few minutes, then manually check
if the data that got historised in the sender was pushed successfully to
the receiver.

sjlongland committed:
[rt] nhaystack: Support truly null `BHRef`.

Presently, if you want a "null" BHRef, the value chosen is a literal
`@null`, which is not intuitive for the user.  It would be better to
have the user just leave a field completely blank.

This however causes parsing errors with the `BHRef` class.

The following commit amends `BHRef` so it can support representing a
"null" `HRef`.  That is, a `HRef` that is not presently set to point to
any given Project Haystack entity.

sjlongland committed:
[rt] nhaystack.driver.history: Add export class.

This implements the core export algorithm used to push history from a
Niagara station's history to an upstream Project Haystack server.  The
class also supports discovery of points (by way of a filtered read) or
(if the server supports it) creating the points as required.

sjlongland committed:
[rt] nhaystack.driver.history.learn: Add discovery classes

This adds the necessary classes for "learning" about points that could
be exported via Project Haystack history uploads.

sjlongland committed:
[wb] nhaystack.ui.history: Add export manager class.

This is the class that provides the management UI for exported
histories.

sjlongland committed:
[rt] nhaystack.driver.BNHaystackServer: Add history discovery.

This adds an action slot which "discovers" histories that can be
exported to an upstream Project Haystack server.

sjlongland committed:
[rt] nhaystack.driver.history.BNHaystackHistoryDeviceExt: Add export descriptor

sjlongland committed:
[rt] nhaystack.driver.BNHaystackServer: Add feature detection.

This adds detection of server features to the core server class so that
it can automatically enable or disable support for optional features
such as CRUD.

sjlongland committed:
Update documentation.

This adds notes on how to configure `nHaystack` for pushing history to an
upstream Project Haystack server.

sjlongland committed:
[rt] nhaystack.driver: Delay first hisWrite after a createRec.

Some implementations of Project Haystack require some time after
creation of an entity to re-load authorisation caches before they'll
permit us to do a `hisWrite` operation.  This usually is no more than a
few seconds, but can vary from server-to-server.

Delaying our first `hisWrite` a little will lead to a better user
experience as it will ensure this re-load can take place before we
attempt the write.

sjlongland committed:
[rt, wb]: Search and replace `ax(Station|HistoryId|SlotPath)`

Replace the prefix on these tags to `n4` to reflect these are in fact
being used with Niagara 4 not Niagara AX.

ci-richard-mcelhinney commented on nhaystack-wb/src/nhaystack/ui/Row.java:
```diff
@@ -306,11 +306,11 @@ private static BWbFieldEditor makeRefFE(
-                if (dict.has("axSlotPath"))
-                    ord = BOrd.make("station:|" + dict.getStr("axSlotPath"));
+                // workaround for weird problem with n4SlotPath
+                if (dict.has("n4SlotPath"))
```
Unfortunately this would be a breaking change for many people that have been using the module for years across Niagara AX and Niagara 4.  I understand the desire to change and update, but this would break too many installations, both tags would need to be accomodated here.

sjlongland replied:
No problems, the change was made because I recall the suggestion made that we should use `n4` as a prefix in preference to `ax`.  I wasn't aware of any production systems using this feature so didn't anticipate breakage.

I've added logic that falls back to using `axSlotPath` in the situation where `n4SlotPath` is absent, so it should handle either case.

ci-richard-mcelhinney commented on nhaystack-wb/src/nhaystack/ui/BSiteRefFE.java:
```diff
@@ -57,7 +57,7 @@ public BSiteRefFE(BHDictEditorGroup group)
         for (int i = 0; i < grid.numRows(); i++)
         {
-            String slotPath = grid.row(i).getStr("axSlotPath");
+            String slotPath = grid.row(i).getStr("n4SlotPath");
```
Breaking change.

sjlongland replied:
Logic changed now: we look for `n4SlotPath`, but if that's missing, we use `axSlotPath` here instead.

ci-richard-mcelhinney replied:
For some reason I'm not seeing the update as you describe it.  I must be looking at something incorrectly.

ci-richard-mcelhinney commented on nhaystack-wb/src/nhaystack/ui/BEquipRefFE.java:
```diff
@@ -85,7 +85,7 @@ private static String findImplicitEquip(BComponent comp, HGrid equipsGrid)
         for (int i = 0; i < equipsGrid.numRows(); i++)
         {
-            String slotPath = equipsGrid.row(i).getStr("axSlotPath");
+            String slotPath = equipsGrid.row(i).getStr("n4SlotPath");
```
Breaking change.

sjlongland replied:
Logic changed now: we look for `n4SlotPath`, but if that's missing, we use `axSlotPath` here instead.

ci-richard-mcelhinney commented on nhaystack-wb/src/nhaystack/ui/BEquipRefFE.java:
```diff
@@ -65,7 +65,7 @@ public BEquipRefFE(BHDictEditorGroup group)
         for (int i = 0; i < equipsGrid.numRows(); i++)
         {
-            String slotPath = equipsGrid.row(i).getStr("axSlotPath");
+            String slotPath = equipsGrid.row(i).getStr("n4SlotPath");
```
Same here, breaking change.

sjlongland replied:
Logic changed now: we look for `n4SlotPath`, but if that's missing, we use `axSlotPath` here instead.

ci-richard-mcelhinney commented on nhaystack-rt/test/nhaystack/test/SupervisorClientTest.java:
```diff
@@ -270,7 +270,7 @@ void verifySupHisRead()
     try
     {
-      client.read("axHistoryId==\"/nhaystack_j1/SineWave1\"");
+      client.read("n4HistoryId==\"/nhaystack_j1/SineWave1\"");
```
As per the other comments, this would also be a breaking change.  We need to maintain the current tags that are supported.  We can look to make a plan to add support for new tags but we can't remove the existing ones.

sjlongland replied:
The code here does a `read` operation to retrieve a `HDict` of the target entity, then does a `hisRead` on it; the latter grid contains the ID of the resolved entity, and the test case asserts these IDs match.

I've added additional checks which queries using `axSlotPath` or `axHistoryId`, then compares the grid from _that_ latter query still matches the ID from the `hisRead` (which we asserted, matched the original `read` response).

ci-richard-mcelhinney commented on nhaystack-rt/test/nhaystack/test/SupervisorClientTest.java:
```diff
@@ -257,7 +257,7 @@ void verifySupHisRead()
 
-    HDict dict = client.read("axSlotPath==\"slot:/Drivers/NiagaraNetwork/nhaystack_j2/points/SineWave2\"");
+    HDict dict = client.read("n4SlotPath==\"slot:/Drivers/NiagaraNetwork/nhaystack_j2/points/SineWave2\"");
     HGrid his = client.hisRead(dict.id(), "today");
```
Breaking change.

ci-richard-mcelhinney commented on nhaystack-rt/src/nhaystack/site/BHSite.java:
```diff
@@ -362,7 +362,7 @@ public HDict generateTags(NHServer server)
         // add misc other tags
         hdb.add("axType", getType().toString());
-        hdb.add("axSlotPath", getSlotPath().toString());
+        hdb.add("n4SlotPath", getSlotPath().toString());
```
Breaking change.

ci-richard-mcelhinney commented on nhaystack-rt/src/nhaystack/site/BHEquip.java:
```diff
@@ -141,7 +141,7 @@ public HDict generateTags(NHServer server)
         // add misc other tags
         hdb.add("axType", getType().toString());
-        hdb.add("axSlotPath", getSlotPath().toString());
+        hdb.add("n4SlotPath", getSlotPath().toString());
```
Breaking change.

ci-richard-mcelhinney commented on nhaystack-rt/src/nhaystack/server/tags/BNSiteRefRelation.java:
```diff
@@ -134,16 +146,4 @@ private static boolean hasSiteTag(Entity entity)
     {
         return entity.tags().contains(ID_SITE);
     }
-
```
@sjlongland did you remove this?  Or was it already removed in another commit?

sjlongland replied:
Not to point the finger at anyone -- it is possible this is a commit that was on the `n4` branch upon which this code is based.

This branch is merging into `n4-development` as it's a major change that should not be merged into a "production" branch without further evaluation, however to ensure trouble-free merging into the `n4` branch later, I've based the branch on `n4`.

```
$ git log nhaystack-rt/src/nhaystack/server/tags/BNSiteRefRelation.java
commit 2ea3672aff905f073b83b76fd99b4464a87773d1
Author: Eric Anderson <esanderson@uwalumni.com>
Date:   Mon Jan 18 21:51:51 2021 -0500

    Fix major performance issue in NSiteRefRelation with get all relations
    
    Also, removed redundancy from TagManager that retrieved all relations
    twice.

commit c65b7cb515475f708bc3d7ea2dea68146c001d9b
Author: Eric Anderson <eric.anderson3@tridium.com>
Date:   Fri Apr 26 00:02:59 2019 -0400

    Added nHaystack version of smart tags for id, equipRef, and siteRef
    Making plain components tagged with site or equip "visible"
    Made equip a frozen tag on BHEquip
    Added spy on Cache
```

ci-richard-mcelhinney commented on nhaystack-rt/src/nhaystack/server/tags/BNSiteRefRelation.java:
@sjlongland I can't tell if you've modified anything in this file.  I suspect not but would like to confirm please?

sjlongland replied:
Looks to be a change made in 2ea3672aff905f073b83b76fd99b4464a87773d1.

ci-richard-mcelhinney commented on nhaystack-rt/src/nhaystack/server/tags/BNEquipRefRelation.java:
Same here, have you modified anything in this file?

sjlongland replied:
2ea3672aff905f073b83b76fd99b4464a87773d1 again.

ci-richard-mcelhinney commented on nhaystack-rt/src/nhaystack/server/TagManager.java:
```diff
@@ -1266,10 +1274,10 @@ private static int getTrendRecordKind(Type trendRecType)
     static final String[] AUTO_GEN_TAGS = {
         "axAnnotated",
-        "axHistoryId",
+        "n4HistoryId",
```
Breaking change.

sjlongland replied:
Yep, I've just put these back, we'll generate both so old clients don't get left behind.

ci-richard-mcelhinney replied:
It's still showing as the axHistoryId as removed, am I reading it wrong?

ci-richard-mcelhinney commented on nhaystack-rt/src/nhaystack/server/TagManager.java:
```diff
@@ -1266,10 +1274,10 @@ private static int getTrendRecordKind(Type trendRecType)
         "axHistoryRef",
         "axPointRef",
-        "axSlotPath", 
+        "n4SlotPath", 
```
Breaking change.

sjlongland replied:
Re-instated.

ci-richard-mcelhinney replied:
It's stil lshowing as the axSlotPath has been removed, am I reading it wrong?

ci-richard-mcelhinney reviewed (changes requested):
@sjlongland there are some breaking changes that need to be addressed.  Also, there seems to be a lot of formatting changes that are creating noise in the PR and I can't tell if they're from your changes or from previous commits.

Is there a way we can reduce this PR to just your changes please?

Commit:
Small tweaks to satisfy linter

Commit:
More tweaks to satisfy linter

sjlongland committed:
[rt] BNHaystackHistoryExport: Restore support reading points by `ax` tags.

Support both `n4Station`/`n4HistoryId` *and* `axStation`/`axHistoryId`
to ensure existing sites are not broken.

New points will be created still with `n4Station`/`n4HistoryId` since
presently only WideSky supports the `createRec` operation as implemented
here and no WideSky sites are presently using nHaystack for history
upload.

sjlongland committed:
[rt] blacklist.txt: Restore `axHistoryId`/`axSlotPath`

Add `n4HistoryId`/`n4SlotPath`, retaining alphabetical ordering.

sjlongland committed:
[rt] NHServerOps: Try `axSlotPath` if `n4SlotPath` is missing.

sjlongland committed:
[rt] Nav: Return both `axSlotPath` and `n4SlotPath`

Return the slot path under both tags for clients that may look for it in
one place or the other.

Long term, we should have some way to tell the server which tag we
intend to expose, but this is the safest "least invasive" approach for
now.

sjlongland committed:
[rt] TagManager: Add both `axSlotPath` and `n4SlotPath`

sjlongland committed:
[rt] BHEquip: Retain `axSlotPath` along side `n4SlotPath`

sjlongland committed:
[rt] BHSite: Retain `axSlotPath`

sjlongland committed:
[rt] BSimpleClientTest: Ensure we haven't broken `axSlotPath` queries.

sjlongland committed:
[rt] BHaystackClientTest: Ensure we don't break the old tags.

sjlongland committed:
[rt] SimpleClientTest: Ensure `axSlotPath` / `axHistoryId` still work

sjlongland committed:
[rt] SupervisorClientTest: Ensure we didn't break the old tags.

sjlongland committed:
[wb] Row: Use `axSlotPath` if `n4SlotPath` is absent.

sjlongland committed:
[wb] BEquipRefFE: Fall back to `axSlotPath` if `n4SlotPath` is absent.

sjlongland committed:
[wb] BSiteRefFE: Fall back to `axSlotPath` if `n4SlotPath` is absent

sjlongland committed:
[rt] TagManager: Re-instate `axSlotPath` / `axHistoryId`.

Missed these ones.

sjlongland committed:
[rt] TagManager: Fix missed semicolon.

sjlongland committed:
[rt] BNHaystackLearnExpHistoriesJob: `TZ_REGIONS` is protected now.

sjlongland committed:
[wb] BEquipRefFE, BSiteRefFE: Import `HRow` from `haystack-java` lib.

ci-richard-mcelhinney commented on nhaystack-rt/src/nhaystack/driver/history/BNHaystackHistoryExport.java:
```diff
@@ -205,9 +205,9 @@ private HRef locatePoint(HClient client)
     BHistoryId id = getHistoryId();
     StringBuilder pointFilter = new StringBuilder();
-    pointFilter.append("point and axStation==");
+    pointFilter.append("point and n4Station==");
```
Whilst I know this is your new code to do the export, I think you should look to achieve backwards compatibility by including the axStation tag in the filter.

ci-richard-mcelhinney commented on nhaystack-rt/src/nhaystack/driver/history/BNHaystackHistoryExport.java:
```diff
@@ -205,9 +205,9 @@ private HRef locatePoint(HClient client)
+    pointFilter.append("point and n4Station==");
     pointFilter.append(HStr.make(Sys.getStation().getStationName()).toZinc());
-    pointFilter.append(" and axHistoryId==");
+    pointFilter.append(" and n4HistoryId==");
```
Same as previous comment about trying to achieve backwards compatibility with existing stations.

ci-richard-mcelhinney commented on nhaystack-rt/src/nhaystack/res/blacklist.txt:
```diff
@@ -1,9 +1,9 @@
 actions
 axAnnotated
-axHistoryId
+n4HistoryId
```
I need to review how this blacklist is used before approving this change and the same a few lines down.  If this is used elsewhere then we probably need to maintain the existing 'ax' tags.

ci-richard-mcelhinney commented on nhaystack-rt/src/nhaystack/server/NHServerOps.java:
```diff
@@ -1508,7 +1508,7 @@ public static BComponent[] getFilterComponents(NHServer server, String filter, S
       for (int i = 0; i < grid.numRows(); i++)
       {
-        HStr slotPath = (HStr) grid.row(i).get("axSlotPath", false);
+        HStr slotPath = (HStr) grid.row(i).get("n4SlotPath", false);
```
I think the axSlotPath tag should be included here for systems that this might be applied to retrospectively.  People won't go back and change their tags on an existing system if they upgrade the module.  So for backwards compatibility it needs to include the axSlotPath tag.

ci-richard-mcelhinney replied:
Ok now I can see it properly it looks like it's handled better.  Thanks.

ci-richard-mcelhinney commented on nhaystack-rt/src/nhaystack/server/Nav.java:
```diff
@@ -256,7 +256,7 @@ private HDict makeCompNavRec(BComponent comp)
             hdb.add("dis", comp.getDisplayName(null));
             hdb.add(AXTYPE, comp.getType().toString());
-            hdb.add("axSlotPath", comp.getSlotPath().toString());
+            hdb.add("n4SlotPath", comp.getSlotPath().toString());
```
Need to account for existing systems that will use the axSlotPath tag.

ci-richard-mcelhinney commented on nhaystack-rt/src/nhaystack/server/TagManager.java:
```diff
@@ -485,7 +485,7 @@ public HDict createComponentTags(BComponent comp)
             hdb.add("axType", comp.getType().toString());
             if (comp.getSlotPath() != null)
-                hdb.add("axSlotPath", comp.getSlotPath().toString());
+                hdb.add("n4SlotPath", comp.getSlotPath().toString());
```
We can't just change how we treat the axSlotPath tag.  There will be existing systems that use it needs to be kept.

ci-richard-mcelhinney commented:
Ok I think the way I'm reviewing the commits is back to front.  Github is showing me the commits in oldest to last, it's very different to bitbucket that we use and still incredibly hard to tell what's going on. 

I'll try and do the review again.  I'm sure you've addressed all my previous comments, but I can't see them.

ci-richard-mcelhinney reviewed (approved):
Thanks @sjlongland for making those changes.  I managed to figure out Github reviews now!  So all good.  Happy to approve and move forwards.
